maro/rl_v3/training/algorithms/ac.py

```diff
@@ -211,20 +211,21 @@ def record(self, exp_element: ExpElement) -> None:
     def get_local_ops_by_name(self, ops_name: str) -> AbsTrainOps:
         return DiscreteActorCriticOps(get_policy_func=self._get_policy_func, **self._params.extract_ops_params())
 
-    def _get_batch(self, agent_name: str) -> TransitionBatch:
-        return self._replay_memory_dict[agent_name].sample(-1)  # Use all entries in the replay memory
-
     async def train_step(self):
-        for agent_name in self._replay_memory_dict:
-            batch = self._ops.preprocess_batch(self._get_batch(agent_name))
-            for _ in range(self._params.grad_iters):
-                batches = [batch] if self._params.data_parallelism == 1 else batch.split(self._params.data_parallelism)
-                critic_grad_list = [self._ops.get_critic_grad(batch) for batch in batches]
-                if isinstance(self._ops, RemoteOps):
-                    critic_grad_list = await asyncio.gather(*critic_grad_list)
-
-                self._ops.update_critic(average_grads(critic_grad_list))
-                actor_grad_list = [self._ops.get_actor_grad(batch) for batch in batches]
-                if isinstance(self._ops, RemoteOps):
-                    actor_grad_list = await asyncio.gather(*actor_grad_list)
-                self._ops.update_actor(average_grads(actor_grad_list))
+        batch_list = []
+        for memory in self._replay_memory_dict.values():
+            batch = self._ops.preprocess_batch(memory.sample(-1))  # Use all entries in the replay memory
```

Comment:
Would it be possible to move preprocess_batch into ``record``?

Reply:
I think it is impossible. `record` only storage a part of experiences, even a single experience entry, but the batch can be processed only after it is completely recorded.

Reply:
I said it wrong. What I meant was move ``preprocess_batch`` out of ``train_step``. We can preprocess the batch for every (env_idx, agent_name) pair and then put them all together. That way we can get rid of the ``for memory in self._replay_memory_dict.values():`` in ``train_step`` and just use the entire batch for training.

Reply:
Sure, I'll add a `_get_batch()` method for AC.

Reply:
Sure, I'll add a `_get_batch()` method for AC.